File: LANGUAGE  PRACTICE/c  practice/queue/queue_by_LL.c
#include<stdio.h>
#include<stdlib.h>
void enqueue(int);
void dequeue();
void display();
void peek();
struct node
{
    int data;
    struct node *next;
};
struct node *rear=NULL;
struct node *front=NULL;
void main()
{
    enqueue(10);
    enqueue(20);
    display();
    dequeue();
    peek();
    dequeue();
    dequeue();
}
void enqueue(int x)
{
    struct node *newnode;
    newnode=(struct node*)malloc(sizeof(struct node));
    newnode->data=x;
    newnode->next=NULL;
    if(rear==NULL&&front==NULL)
    {
        rear=front=newnode;
    }
    else{
        rear->next=newnode;
        rear=newnode;
    }
}
void dequeue()
{
    if(rear==NULL&&front==NULL)
    {
        printf(" \nqueue is empty");
    }
    else
    {
        struct node *temp;
        temp=front;
        front=front->next;
        if(front == NULL) // If the queue becomes empty after dequeuing
        {
            rear = NULL;
            printf(" \nqueue is empty");
        }
        free(temp);
    }
}
void display()
{
    if(rear==NULL&&front==NULL)
    {
        printf(" \nqueue is empty");
    }
    else 
    {
        struct node *temp;
        temp=front;
        while(temp!=NULL)
        {
            printf("%d\t",temp->data);
            temp=temp->next;
        }
    }
}
void peek()
{
    if(rear==NULL&&front==NULL)
    {
        printf(" \nqueue is empty");
    }
    else{
        printf("\n peek is %d",front->data);
    }
}

File: DSA/soc python/queue/queue_by_LL.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class QueueLinkedList:
    def __init__(self):
        self.front = self.rear = None

    def enqueue(self, data):
        new_node = Node(data)
        if not self.rear:
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node

    def dequeue(self):
        if not self.front:
            return "Queue is empty"
        result = self.front.data
        self.front = self.front.next
        if not self.front:
            self.rear = None
        return result

    def display(self):
        temp = self.front
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("None")
print("Queue using Linked List:")
q = QueueLinkedList()
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.display()           # 1 -> 2 -> 3 -> None
print(q.dequeue())    # 1
q.display()           # 2 -> 3 -> None
